solidity/contracts/vortex/VortexBurner.sol

```diff
@@ -0,0 +1,558 @@
+    ITokenGovernance public immutable _govTokenGovernance;
+
+    // the address of the vortex stats contract
+    IVortexStats private immutable _stats;
+
+    // the percentage of the converted network tokens to be sent to the caller of the vortex (in units of PPM)
+    uint32 private _burnIncentiveFee;
+
+    // the maximum incentive fee to be sent to the caller of the vortex
+    uint256 private _maxBurnIncentiveFeeAmount;
+
+    /**
+     * @dev triggered when the burn incentive fee has been changed
+     *
+     * @param prevBurnIncentiveFee the previous burn incentive fee (in units of PPM)
+     * @param newBurnIncentiveFee the new burn incentive fee (in units of PPM)
+     * @param prevMaxBurnIncentiveFeeAmount the previous maximum burn incentive fee
+     * @param newMaxBurnIncentiveFeeAmount the new maximum burn incentive fee
+     */
+    event BurnIncentiveFeeUpdated(
+        uint32 prevBurnIncentiveFee,
+        uint32 newBurnIncentiveFee,
+        uint256 prevMaxBurnIncentiveFeeAmount,
+        uint256 newMaxBurnIncentiveFeeAmount
+    );
+
+    /**
+     * @dev triggered after a completed vortex
+     *
+     * @param tokens the converted tokens
+     * @param amounts the amounts of the converted tokens
+     * @param conversionAmounts the network token amounts the tokens were converted to
+     * @param totalBurnedAmount the total burned amount in this vortex
+     */
+    event Burned(IERC20[] tokens, uint256[] amounts, uint256[] conversionAmounts, uint256 totalBurnedAmount);
+
+    /**
+     * @dev initializes a new VortexContract contract
+     *
+     * @param networkToken the address of the network token
+     * @param govTokenGovernance the address of the governance token security module
+     * @param stats the address of the vortex stats contract
+     * @param burnIncentiveFee the percentage of the converted network tokens to be sent to the caller of the vortex (in units of PPM)
+     * @param maxBurnIncentiveFeeAmount the maximum incentive fee to be sent to the caller of the vortex
+     * @param registry the address of the contract registry
+     */
+    constructor(
+        IERC20 networkToken,
+        ITokenGovernance govTokenGovernance,
+        IVortexStats stats,
+        uint32 burnIncentiveFee,
+        uint256 maxBurnIncentiveFeeAmount,
+        IContractRegistry registry
+    )
+        public
+        validAddress(address(networkToken))
+        validAddress(address(govTokenGovernance))
+        validAddress(address(stats))
+        validFee(burnIncentiveFee)
+        ContractRegistryClient(registry)
+    {
+        _networkToken = networkToken;
+        _govTokenGovernance = govTokenGovernance;
+        _govToken = govTokenGovernance.token();
+        _stats = stats;
+
+        _burnIncentiveFee = burnIncentiveFee;
+        _maxBurnIncentiveFeeAmount = maxBurnIncentiveFeeAmount;
+    }
+
+    // ensures that the fee is valid
+    modifier validFee(uint32 fee) {
```

Comment:
Already implemented in parent contract `Utils`.

Reply:
It seems that it's only in `NetworkSettings`. I'll move it to `Utils`

Reply:
Watch out! It's implemented in NetworkSettings separately because the network fee is allowed to be zero, while we generally don't allow zero portions elsewhere. So the modifier in the NetworkSettings contract is different from the one in Utils.

Reply:
That's actually fine, because the burning incentive fee can zero as well. I think that, in any case, this should be the use case and the difference between `validPortion` and `validFee`, like in both of these cases.